### PR TITLE
Generative Replay

### LINKED ISSUE #927: Implementation of a Generative Replay Strategy
### Discussed in https://github.com/ContinualAI/avalanche/discussions/414

Introduce a plugin-based implementation of Generative Replay, see [the original paper](https://arxiv.org/abs/1705.08690) for details.

### PR DESCRIPTION
**Pull request by travela**
closes #927 

This is adds a `GenerativeReplayPlugin` and a `GenerativeReplay` strategy to the library with which one can train models according to the vanilla Generative Replay algorithm. One can either apply it to a generator model alone or a pair consisting of a classifier and a generator.

In particular I have added two usage examples where we train two models on the splitMNIST scenario, i.e.:

- the SimpleMLP model  in `examples/generative_replay_splitMNIST.py` and
-  a VAE model in `examples/generative_replay_MNIST_generator.py`

### CODE REVIEW

**Comment by AntonioCarta** [comment]
Hey @travela thanks for your contribution. It looks solid in general, I just left a couple of minor comments.
Did you reproduce the results on Split MNIST? It would be ideal to have a script to add to the [reproducible-cl](https://github.com/ContinualAI/reproducible-continual-learning) repository, as we do for the other strategies.

**Comment by travela** [comment]
> Hey @travela thanks for your contribution. It looks solid in general, I just left a couple of minor comments. Did you reproduce the results on Split MNIST? It would be ideal to have a script to add to the [reproducible-cl](https://github.com/ContinualAI/reproducible-continual-learning) repository, as we do for the other strategies.

Hi @AntonioCarta, great thanks! I don't seem to be able to see your comments when looking in the "Files changed" tab. Did you publish them or could it be that they are still pending?  

And yes, I reproduced the results on the 10 classes splitMNIST scenario in `generative_replay_splitMNIST.py`. I will look into the reproducible-cl repo and try to convert my example into a test script that can be added.

**Review by AntonioCarta — CHANGES_REQUESTED** [review]

File: avalanche/training/plugins/generative_replay.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        if self.untrained_solver:
+            # The solver needs to be trained before labelling generated data and
+            # the generator needs to be trained before we can sample.
+            self.untrained_solver = False
+            return
+
+        # Sample data from generator
+        memory = self.generator.generate(
+            len(strategy.adapted_dataset) *
+            (strategy.experience.current_experience)).to(strategy.device)
```
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:106]: Good question! There are two approaches (as mentioned [here](https://www.nature.com/articles/s41467-020-17866-2#Sec19)): either we generate all data needed for the experience beforehand and then start the training, or we save a copy of the "old" generator and generate our data on demand. I chose the former approach also because this way I could stay in line with the existing Replay strategy, where we update the strategy.data_loader once before each experience. 

Do you you think the other method would be more efficient? How would the implementation of an additional dataloader roughly look like? Since the memory variable is only used temporarily and it is then passed to the ReplayDataloader in the next step, I was hoping that we are indeed already making use of the dataloader properties and do not clog RAM too much (but I am not entirely sure about this).

[the remark this answers is not in the document]
Comment (AntonioCarta) [reply avalanche/training/plugins/generative_replay.py:106]: Both are reasonable solutions. If you want to use this one, you should:
- batch the results of the Generator. In general you want minibatch size < data size, because the entire dataset may be too large to generate in one step.
- for the same reasons, unless the dataset is small, it's better to keep it in cpu and move only the minibatch to the gpu when needed.

If you don't do this you will get an out-of-memory error.

[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:106]: I see. How would the first point tie in with the ReplayDataLoader? I could generate the data one mini-batch at a time to make it more resistant for larger datasets, but then I would still have to obtain an AvalancheDataset (ie. to concatenate them again?). This is because ReplayDataLoader expects an AvalancheDataset and not some kind of Dataloader, if I understand it correctly?

[the remark this answers is not in the document]
Comment (AntonioCarta) [reply avalanche/training/plugins/generative_replay.py:106]: > This is because ReplayDataLoader expects an AvalancheDataset and not some kind of Dataloader, if I understand it correctly?

Exactly. I would make another dataloader similar to ReplayDataLoader that accepts a dataset (the current data), and an iterator (data generator).

[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:106]: I definitely like the idea of using a Python Generator. But that would then correspond to the second of the two approaches I mentioned above, as when yielding the next batch we need access to the old version of the generator. I was trying to avoid having to store an additional model as well as trying to stick to the existing ReplayDataLoader as I understood it to be designed for any "rehearsal/replay strategies". 

I am aiming to update the pull request by the beginning of next week.

[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py]: Hi! I just merged my changes ([bba78b9](https://github.com/ContinualAI/avalanche/pull/931/commits/bba78b919f2b6fff66c0b69147c5e914fc9ab9bc)). As mentioned in my last comment I now opted for the "on-demand" replay data generation, where we store the old generator+model and create replay data before each training iteration. This way I got around using the whole ReplayDataLoader syntax and instead just extend the current mini-batch with newly generated replay data.

Accuracy on splitMNIST stayed the same.

File: avalanche/training/plugins/generative_replay.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    training the solver of a scholar model, we also train the generator on the 
+    data of the current experience.
+    """
+
+    def after_training_exp(self, strategy: "SupervisedTemplate", **kwargs):
+        """
+        The training method expects an Experience object 
+        with a 'dataset' parameter.
+        """
+        strategy.plugins[1].generator_strategy.train(strategy.experience)
```
Comment (ggraffieti) [inline avalanche/training/plugins/generative_replay.py:157]: How do you know that the `GenerativeReplayPlugin` has index 1 in the plugins list? 
Generally you cannot make this assumption!

You can cycle throughout all the plugins and select the one you interested with using something like 
```python
for plugin in strategy.plugins:
    if type(plugin) is GenerativeReplayPlugin:
        plugin.generator_strategy.train(strategy.experience) 
```

File: avalanche/training/supervised/strategy_wrappers.py [hunk 3]
```diff
@@ -276,6 +281,187 @@ def __init__(
...
+        train_mb_size: int = 1,
+        train_epochs: int = 1,
+        eval_mb_size: int = None,
+        device=None,
+        plugins: Optional[List[SupervisedPlugin]] = None,
+        evaluator: EvaluationPlugin = default_evaluator,
+        eval_every=-1,
+        generator_strategy: BaseTemplate = None,
+        **base_kwargs
+    ):
```
Comment (ggraffieti) [inline avalanche/training/supervised/strategy_wrappers.py:317]: Really like the idea of having 2 different strategies, one for the classifier (this one) and one for the generative model (passed as an argument). This allows a great generalization.

**Review by ggraffieti — CHANGES_REQUESTED** [review]
Very nice work! 
I added some comments and little requests, but this seems a great starting point!
I labeled this review as _request changes_ only to follow in details the next steps, but the work to be done is fairly low. 
As a general comment, many classes are named quite generally (Generator, VAE, VAEEncoder..) but lack this generality in the implementation (e.g. a CNN VAE). I'd suggest to call them in different ways (e.g. SimpleVAE or MlpVAE) in order to don't have the need, in the future, to rename them losing compatibility with old code.

File: avalanche/models/generator.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            out_dim = hidden_size[i+1]
+            q.append(("Linear_%d" % i, nn.Linear(in_dim, out_dim)))
+            if (i < len(hidden_size)-2) or ((i == len(hidden_size) - 2)
+                                            and (last_activation)):
+                q.append(("BatchNorm_%d" % i, nn.BatchNorm1d(out_dim)))
+                q.append(("ReLU_%d" % i, nn.ReLU(inplace=True)))
+        self.mlp = nn.Sequential(OrderedDict(q))
+
+    def forward(self, x):
+        return self.mlp(x)
```
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/models/generator.py:86]: Agreed! I'm happy to add it to the utils. (commit c9f5a230b8f2aaf72531eb710564b316ec4dc121)

File: avalanche/models/generator.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    A base abstract class for generators
+    """
+
+    @abstractmethod
+    def generate(self, batch_size=None):
+        """
+        Lets the generator sample random samples.
+        Output is either a single sample or, if provided,
+        a batch of samples of size "batch_size" 
+        """
```
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/models/generator.py:39]: I think in any case it won't hurt to have it as an optional parameter in the abstract method. I've added _condition_ as an input which simply defaults to None. (commit d9c1f1c234f3021480f4ebda2b7b5ebcbd21f815)

File: avalanche/training/plugins/generative_replay.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            self.generator = None
+        self.untrained_solver = untrained_solver
+        self.model_is_generator = False
+
+    def before_training(self, strategy: "SupervisedTemplate", *args, **kwargs):
+        """Checks whether we are using a user defined external generator 
+        or we use the strategy's model as the generator. 
+        If the generator is None after initialization 
+        we assume that strategy.model is the generator."""
+        if not self.generator_strategy:
```
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:85]: Yes exactly: if no generative strategy is passed, we assume that the strategy our plugin was added to already has a model which is generative. E.g. this allows us to easily train a generator with generative replay by simply adding the plugin. In that case there would be no classifier. (Another scenario is that the classifier and generator are combined in a single model, as in this paper about [Generative replay with feedback connections](https://arxiv.org/abs/1809.10635)).

I agree that it can be confusing (a result of trying to combine all scenarios in a single plugin). I added another line in the doc string (4f5246f65d499df2e9015bbdd4d9fa2122c1f378) to refer to the example of training a generator, hoping this would make it clearer. Or do you have another suggestion?

File: avalanche/training/plugins/generative_replay.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+                 untrained_solver: bool = True):
+        '''
+        Init.
+        '''
+        super().__init__()
+        self.mem_size = mem_size
+        self.batch_size = batch_size
+        self.batch_size_mem = batch_size_mem
+        self.task_balanced_dataloader = task_balanced_dataloader
+        self.generator_strategy = generator
```
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:72]: This is indeed not clear. It's a relict from an older version.

We expect a generative strategy as an input, I have updated the variable names accordingly (3f096b10f6a4969aa7c3d49e545810e08dfa2121).

File: avalanche/training/plugins/generative_replay.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    :param task_balanced_dataloader: if True, buffer data loaders will be
+            task-balanced, otherwise it will create a single dataloader for the
+            buffer samples.
+    :param untrained_solver: if True we assume this is the beginning of 
+        a continual learning task and add replay data only from the second 
+        experience onwards, otherwise we sample and add generative replay data
+        before training the first experience. Default to True.
+    """
+
+    def __init__(self, generator=None, mem_size: int = 200,
```
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:59]: Done! (03150d1b324729ce287a47414ec5b638a5070fc6)

File: avalanche/training/plugins/generative_replay.py (hunk 2 above)
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:157]: Perfect, thank you! Looks a lot more robust now :+1:

File: avalanche/training/plugins/generative_replay.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        each training iteration.
+        """
+        if self.untrained_solver:
+            # The solver needs to be trained before labelling generated data and
+            # the generator needs to be trained before we can sample.
+            return
+        # extend X with replay data
+        replay = self.old_generator.generate(
+            len(strategy.mbatch[0]) * (strategy.experience.current_experience)
+            ).to(strategy.device)
```
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py:128]: I actually got inspired to do this from this [intro_to_generative_replay notebook](https://github.com/ContinualAI/colab/blob/master/notebooks/intro_to_generative_replay.ipynb) from the Avalanche/colab repo. In the literature I usually see authors employ a weighted loss instead  to achieve a similar effect (e.g. as described here in [equation 3](https://www.nature.com/articles/s41467-020-17866-2#Sec17)). It could be due to the VAE and classifier I use, but in my case the increasing generated data fared a few percentage points higher in accuracy than the weighted loss approach.

File: avalanche/training/plugins/generative_replay.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (travela) [reply avalanche/training/plugins/generative_replay.py]: Good point. I have actually implemented an alternative [solution](https://github.com/travela/avalanche/tree/weighted_loss) where we use a weighted loss to gradually reduce the importance of a new experience/class as the total number grows. As mentioned below, this solution yielded a lower accuracy (in my particular example). However not using any of the two methods yielded much lower accuracy, suggesting that at least one of the two is necessary. 

When implementing the weighted loss I had to overwrite the criterions of both, the model and the generator, which maybe makes it less intuitive, but I guess proper documentation will take care of that.

I was thinking of adding the weighted loss as an option for the user. Maybe I could make it the default option and offer the increasing minibatch size as an alternative (for simple cases like MNIST)?

**Comment by travela** [comment]
Ciao @ggraffieti! Thanks a lot for the detailed review and the helpful suggestions. I added all requested changes just now.

File: avalanche/training/plugins/generative_replay.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AntonioCarta) [reply avalanche/training/plugins/generative_replay.py]: Makes sense. Consider that the simple replay in Avalanche does not change the batch size, so it would be better to have consistent behavior here.

File: avalanche/training/plugins/generative_replay.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ggraffieti) [reply avalanche/training/plugins/generative_replay.py:69]: No it's fine, I read the code sequentially, so I didn't have a complete idea until the end. 
I still believe is a bit confusing, but I don't have a better idea at the moment. 
I really like the generality of the strategy, and overall the pros of this implementation greatly outshine a bit of confusion in this part :wink:

**Comment by AntonioCarta** [comment]
@ggraffieti are there any other changes that you want on this PR?

**Comment by AntonioCarta** [comment]
I checked the test manually and it seems to work. I'm merging this, and we can investigate the CI problems separately.

@travela thanks for your contribution and please remember to push the reproducibility script to the reproducibility-repo.
